src/CircleEditor/CircleEditorProvider.ts

```diff
@@ -0,0 +1,288 @@
+      _token: vscode.CancellationToken): Promise<CircleEditorDocument> {
+    const document: CircleEditorDocument = await CircleEditorDocument.create(uri);
+
+    const listeners: vscode.Disposable[] = [];
+
+    listeners.push(document.onDidChangeDocument((e) => {
+      // Tell VS Code that the document has been edited by the use.
+      this._onDidChangeCustomDocument.fire({
+        document,
+        ...e,
+      });
+    }));
+
+    listeners.push(document.onDidChangeContent((e) => {
+      // Update all webviews when the document changes
+      for (const webviewPanel of this.webviews.get(document.uri)) {
+        this.postMessage(webviewPanel, e);
+      }
+    }));
+
+    document.onDidDispose(() => disposeAll(listeners));
+    return document;
+  }
+
+  resolveCustomEditor(
+      document: CircleEditorDocument, webviewPanel: vscode.WebviewPanel,
+      _token: vscode.CancellationToken): void|Thenable<void> {
+    this.webviews.add(document.uri, webviewPanel);
+
+    // Setup initial content for the webview
+    webviewPanel.webview.options = {
+      enableScripts: true,
+    };
+    webviewPanel.webview.html = this.getHtmlForWebview(webviewPanel.webview);
+
+    webviewPanel.webview.onDidReceiveMessage((e) => this.onMessage(document, e));
+  }
+
+  saveCustomDocument(document: CircleEditorDocument, cancellation: vscode.CancellationToken):
+      Thenable<void> {
+    return document.save(cancellation);
+  }
+  saveCustomDocumentAs(
+      document: CircleEditorDocument, destination: vscode.Uri,
+      cancellation: vscode.CancellationToken): Thenable<void> {
+    return document.saveAs(destination, cancellation);
+  }
+  revertCustomDocument(document: CircleEditorDocument, cancellation: vscode.CancellationToken):
+      Thenable<void> {
+    return document.revert(cancellation);
+  }
+  backupCustomDocument(
+      document: CircleEditorDocument, context: vscode.CustomDocumentBackupContext,
+      cancellation: vscode.CancellationToken): Thenable<vscode.CustomDocumentBackup> {
+    return document.backup(context.destination, cancellation);
+  }
+
+  private postMessage(panel: vscode.WebviewPanel, body: any): void {
+    panel.webview.postMessage(body);
+  }
+
+  private onMessage(document: CircleEditorDocument, message: any) {
+    switch (message.command) {
+      case MessageDefs.alert:
+        Balloon.error(message.text, false);
+        return;
+      case MessageDefs.request:
+        this.handleRequest(document, message.url, message.encoding);
+        return;
+      case MessageDefs.pageloaded:
+        return;
+      case MessageDefs.loadmodel:
+        document.sendModel(parseInt(message.offset));
+        return;
+      case MessageDefs.finishload:
+        return;
+      case MessageDefs.selection:
+        return;
+      case MessageDefs.edit:
+        document.makeEdit(message);
+        return;
+        // TODO: add MessageDefs and appropriate function to handle this request
```

Comment:
```suggestion
      default:
        // TODO: add MessageDefs and appropriate function to handle this request
        return;
```

Reply:
Do we have to add that comment under default?